Implementing IOptionMonad<T, Self> from the dotNext project in a type in another project is impossible

**Version Used**:

    Visual Studio 2022 Professional 17.6.4 targeting .NET 6.0.
    Visual Studio 2022 Professional 17.6.4 targeting .NET 7.0.
    Visual Studio 2022 Professional 17.7.0 Preview 2.0 targeting .NET 8.0.

**Steps to Reproduce**:

1. In a Visual Studio 2022 project (e.g., library or console) reference latest version of the dotNext NuGet package. You can target a wide variety of .NET versions or SDKs, including the preview latest.
2. Try to implement IOptionMonad<T, Self>; e.g., in a readonly struct.
3. In .NET 6.0, you will need to opt in for preview features, which is not a smooth task (EnablePreviewFeatures, RequiresPreviewFeaturesAttribute, and if none works then disabling CA2252).
4. Ask Visual Studio to fill a draft implementation of the interface for you (the one that throws NotImplementedException from every method).

**Expected Behavior**:

The program should compile without errors.

**Actual Behavior**:

I get these three compilation errors:
    - error CS0570: 'IOptionMonad<T, TSelf>.explicit operator T(in TSelf)' is not supported by the language
    - error CS0570: 'IOptionMonad<T, TSelf>.operator true(in TSelf)' is not supported by the language
    - error CS0570: 'IOptionMonad<T, TSelf>.operator false(in TSelf)' is not supported by the language
The interface is unusable outside the project it was implemented at in the first place.
If I include the source code for IOptionMonad<T, TSelf> in my own project (and only that one, not any of the other interfaces IOptionMonad<T, TSelf> itself inherits from), then the errors disappear.

Thank you very much.

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceUserDefinedOperatorSymbolBase.cs
@@ -233,7 +233,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
                 out arglistToken,
                 allowRefOrOut: true,
                 allowThis: false,
-                addRefReadOnlyModifier: false,
+                addRefReadOnlyModifier: IsVirtual || IsAbstract,
                 diagnostics: diagnostics).Cast<SourceParameterSymbol, ParameterSymbol>();
 
             if (arglistToken.Kind() == SyntaxKind.ArgListKeyword)

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/DefaultInterfaceImplementationTests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/DefaultInterfaceImplementationTests.cs
@@ -68992,5 +68992,161 @@ public struct S : MyEnumerable
 }
 """);
         }
+
+        [Fact]
+        public void InParameters_01()
+        {
+            var source1 =
+@"
+public interface I1
+{
+    public static I1 operator +(in I1 x)
+    {
+        return x;
+    }
+
+    public static I1 operator -(in I1 x)
+    {
+        return x;
+    }
+
+    public static I1 operator !(in I1 x)
+    {
+        return x;
+    }
+
+    public static I1 operator ~(in I1 x)
+    {
+        return x;
+    }
+
+    public static I1 operator ++(in I1 x)
+    {
+        return x;
+    }
+
+    public static I1 operator --(in I1 x)
+    {
+        return x;
+    }
+
+    public static bool operator true(in I1 x)
+    {
+        return true;
+    }
+
+    public static bool operator false(in I1 x)
+    {
+        return false;
+    }
+
+    public static I1 operator +(in I1 x, in I1 y)
+    {
+        return x;
+    }
+
+    public static I1 operator -(in I1 x, in I1 y)
+    {
+        return x;
+    }
+
+    public static I1 operator *(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""*"");
+        return x;
+    }
+
+    public static I1 operator /(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""/"");
+        return x;
+    }
+
+    public static I1 operator %(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""%"");
+        return x;
+    }
+
+    public static I1 operator &(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""&"");
+        return x;
+    }
+
+    public static I1 operator |(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""|"");
+        return x;
+    }
+
+    public static I1 operator ^(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""^"");
+        return x;
+    }
+
+    public static I1 operator <<(in I1 x, in int y)
+    {
+        System.Console.WriteLine(""<<"");
+        return x;
+    }
+
+    public static I1 operator >>(in I1 x, in int y)
+    {
+        System.Console.WriteLine("">>"");
+        return x;
+    }
+
+    public static I1 operator >(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine("">"");
+        return x;
+    }
+
+    public static I1 operator <(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""<"");
+        return x;
+    }
+
+    public static I1 operator >=(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine("">="");
+        return x;
+    }
+
+    public static I1 operator <=(in I1 x, in I1 y)
+    {
+        System.Console.WriteLine(""<="");
+        return x;
+    }
+
+    public static I1 operator >>>(in I1 x, in int y)
+    {
+        System.Console.WriteLine("">>>"");
+        return x;
+    }
+}
+";
+
+            var compilation1 = CreateCompilation(source1, options: TestOptions.DebugDll,
+                                                 targetFramework: TargetFramework.NetCoreApp,
+                                                 parseOptions: TestOptions.RegularPreview);
+
+            CompileAndVerify(compilation1, sourceSymbolValidator: validate, symbolValidator: validate, verify: VerifyOnMonoOrCoreClr).VerifyDiagnostics();
+
+            void validate(ModuleSymbol module)
+            {
+                foreach (var m01 in module.GlobalNamespace.GetTypeMember("I1").GetMembers().OfType<MethodSymbol>())
+                {
+                    foreach (var p in m01.Parameters)
+                    {
+                        Assert.Equal(RefKind.In, p.RefKind);
+                        Assert.Empty(p.RefCustomModifiers);
+                    }
+                }
+            }
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs
@@ -33544,5 +33544,301 @@ public class C5 : I1<C5>
                     Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1)
                 );
         }
+
+        [WorkItem("https://github.com/dotnet/roslyn/issues/68767")]
+        [Theory]
+        [InlineData("I1", "+", "(in I1 x)")]
+        [InlineData("I1", "-", "(in I1 x)")]
+        [InlineData("I1", "!", "(in I1 x)")]
+        [InlineData("I1", "~", "(in I1 x)")]
+        [InlineData("I1", "++", "(in I1 x)")]
+        [InlineData("I1", "--", "(in I1 x)")]
+        [InlineData("I1", "+", "(in I1 x, I1 y)")]
+        [InlineData("I1", "-", "(in I1 x, I1 y)")]
+        [InlineData("I1", "*", "(in I1 x, I1 y)")]
+        [InlineData("I1", "/", "(in I1 x, I1 y)")]
+        [InlineData("I1", "%", "(in I1 x, I1 y)")]
+        [InlineData("I1", "&", "(in I1 x, I1 y)")]
+        [InlineData("I1", "|", "(in I1 x, I1 y)")]
+        [InlineData("I1", "^", "(in I1 x, I1 y)")]
+        [InlineData("I1", "<<", "(in I1 x, int y)")]
+        [InlineData("I1", ">>", "(in I1 x, int y)")]
+        [InlineData("I1", ">>>", "(in I1 x, int y)")]
+        [InlineData("I1", "checked -", "(in I1 x)")]
+        [InlineData("I1", "checked ++", "(in I1 x)")]
+        [InlineData("I1", "checked --", "(in I1 x)")]
+        [InlineData("I1", "checked +", "(in I1 x, I1 y)")]
+        [InlineData("I1", "checked -", "(in I1 x, I1 y)")]
+        [InlineData("I1", "checked *", "(in I1 x, I1 y)")]
+        [InlineData("I1", "checked /", "(in I1 x, I1 y)")]
+        public void InParameters_01(string type, string op, string paramList)
+        {
+            string source2 = InParameter_01_Source2(type, op, paramList);
+            InParameter_01_Validate(InParameter_01_Source1_Abstract(type, op, paramList), source2);
+            InParameter_01_Validate(InParameter_01_Source1_Virtual(type, op, paramList), source2);
+        }
+
+        private static string InParameter_01_Source1_Abstract(string type, string op, string paramList)
+        {
+            var source1 =
+@"
+public partial interface I1
+{
+    abstract static " + type + " operator " + op + " " + paramList + @";
+}
+";
+
+            if (op.StartsWith("checked "))
+            {
+                source1 +=
+@"
+partial interface I1
+{
+    abstract static " + type + " operator " + op.Substring(8) + " " + paramList + @";
+}
+";
+            }
+
+            return source1;
+        }
+
+        private static string InParameter_01_Source1_Virtual(string type, string op, string paramList)
+        {
+            var source1 =
+@"
+public partial interface I1
+{
+    virtual static " + type + " operator " + op + " " + paramList + @" => throw null;
+}
+";
+
+            if (op.StartsWith("checked "))
+            {
+                source1 +=
+@"
+partial interface I1
+{
+    virtual static " + type + " operator " + op.Substring(8) + " " + paramList + @" => throw null;
+}
+";
+            }
+
+            return source1;
+        }
+
+        private static string InParameter_01_Source2(string type, string op, string paramList)
+        {
+            var source2 =
+@"
+partial class C1 : I1
+{
+    static " + type + " I1.operator " + op + " " + paramList + @" => throw null;
+}
+";
+
+            if (op.StartsWith("checked "))
+            {
+                source2 +=
+@"
+partial class C1
+{
+    static " + type + " I1.operator " + op.Substring(8) + " " + paramList + @" => throw null;
+}
+";
+            }
+
+            return source2;
+        }
+
+        private void InParameter_01_Validate(string source1, string source2)
+        {
+            var compilation1 = CreateCompilation(source1 + source2, options: TestOptions.DebugDll,
+                                                 targetFramework: _supportingFramework);
+
+            CompileAndVerify(compilation1, sourceSymbolValidator: validate, symbolValidator: validate, verify: Verification.Skipped).VerifyDiagnostics();
+
+            var compilation2 = CreateCompilation(source2, options: TestOptions.DebugDll, references: new[] { compilation1.ToMetadataReference() },
+                                                 targetFramework: _supportingFramework);
+
+            CompileAndVerify(compilation2, verify: Verification.Skipped).VerifyDiagnostics();
+
+            var compilation3 = CreateCompilation(source2, options: TestOptions.DebugDll, references: new[] { compilation1.EmitToImageReference() },
+                                                 targetFramework: _supportingFramework);
+
+            CompileAndVerify(compilation3, verify: Verification.Skipped).VerifyDiagnostics();
+
+            void validate(ModuleSymbol module)
+            {
+                foreach (var m01 in module.GlobalNamespace.GetTypeMember("I1").GetMembers().OfType<MethodSymbol>())
+                {
+                    var modifier = m01.Parameters.First().RefCustomModifiers.Single();
+                    Assert.Equal("System.Runtime.InteropServices.InAttribute", modifier.Modifier.ToTestDisplayString());
+                    Assert.False(modifier.IsOptional);
+
+                    if (m01.ParameterCount > 1)
+                    {
+                        Assert.Empty(m01.Parameters.Last().RefCustomModifiers);
+                    }
+                }
+            }
+        }
+
+        [WorkItem("https://github.com/dotnet/roslyn/issues/68767")]
+        [Theory]
+        [InlineData("I1", "+", "(I1 x, in I1 y)")]
+        [InlineData("I1", "-", "(I1 x, in I1 y)")]
+        [InlineData("I1", "*", "(I1 x, in I1 y)")]
+        [InlineData("I1", "/", "(I1 x, in I1 y)")]
+        [InlineData("I1", "%", "(I1 x, in I1 y)")]
+        [InlineData("I1", "&", "(I1 x, in I1 y)")]
+        [InlineData("I1", "|", "(I1 x, in I1 y)")]
+        [InlineData("I1", "^", "(I1 x, in I1 y)")]
+        [InlineData("I1", "<<", "(I1 x, in int y)")]
+        [InlineData("I1", ">>", "(I1 x, in int y)")]
+        [InlineData("I1", ">>>", "(I1 x, in int y)")]
+        [InlineData("I1", "checked +", "(I1 x, in I1 y)")]
+        [InlineData("I1", "checked -", "(I1 x, in I1 y)")]
+        [InlineData("I1", "checked *", "(I1 x, in I1 y)")]
+        [InlineData("I1", "checked /", "(I1 x, in I1 y)")]
+        public void InParameters_02(string type, string op, string paramList)
+        {
+            string source2 = InParameter_01_Source2(type, op, paramList);
+            InParameter_02_Validate(InParameter_01_Source1_Abstract(type, op, paramList), source2);
+            InParameter_02_Validate(InParameter_01_Source1_Virtual(type, op, paramList), source2);
+        }
+
+        private void InParameter_02_Validate(string source1, string source2)
+        {
+            var compilation1 = CreateCompilation(source1 + source2, options: TestOptions.DebugDll,
+                                                 targetFramework: _supportingFramework);
+
+            CompileAndVerify(compilation1, sourceSymbolValidator: validate, symbolValidator: validate, verify: Verification.Skipped).VerifyDiagnostics();
+
+            var compilation2 = CreateCompilation(source2, options: TestOptions.DebugDll, references: new[] { compilation1.ToMetadataReference() },
+                                                 targetFramework: _supportingFramework);
+
+            CompileAndVerify(compilation2, verify: Verification.Skipped).VerifyDiagnostics();
+
+            var compilation3 = CreateCompilation(source2, options: TestOptions.DebugDll, references: new[] { compilation1.EmitToImageReference() },
+                                                 targetFramework: _supportingFramework);
+
+            CompileAndVerify(compilation3, verify: Verification.Skipped).VerifyDiagnostics();
+
+            void validate(ModuleSymbol module)
+            {
+                foreach (var m01 in module.GlobalNamespace.GetTypeMember("I1").GetMembers().OfType<MethodSymbol>())
+                {
+                    var modifier = m01.Parameters.Last().RefCustomModifiers.Single();
+                    Assert.Equal("System.Runtime.InteropServices.InAttribute", modifier.Modifier.ToTestDisplayString());
+                    Assert.False(modifier.IsOptional);
+
+                    Assert.Empty(m01.Parameters.First().RefCustomModifiers);
+                }
+            }
+        }
+
+        [WorkItem("https://github.com/dotnet/roslyn/issues/68767")]
+        [Theory]
+        [InlineData("bool", "true", "false", "(in I1<T> x)")]
+        [InlineData("I1<T>", ">", "<", "(in I1<T> x, I1<T> y)")]
+        [InlineData("I1<T>", ">=", "<=", "(in I1<T> x, I1<T> y)")]
+        [InlineData("T", "==", "!=", "(in T x, T y)")]
+        public void InParameters_03(string type, string op1, string op2, string paramList)
+        {
+            string source2 = InParameter_03_Source2(type, op1, op2, paramList);
+            InParameter_01_Validate(InParameter_03_Source1_Abstract(type, op1, op2, paramList), source2);
+            InParameter_01_Validate(InParameter_03_Source1_Virtual(type, op1, op2, paramList), source2);
+        }
+
+        private static string InParameter_03_Source1_Abstract(string type, string op1, string op2, string paramList)
+        {
+            var source1 =
+@"
+public partial interface I1<T> where T : I1<T>
+{
+    abstract static " + type + " operator " + op1 + " " + paramList + @";
+    abstract static " + type + " operator " + op2 + " " + paramList + @";
+}
+";
+
+            return source1;
+        }
+
+        private static string InParameter_03_Source1_Virtual(string type, string op1, string op2, string paramList)
+        {
+            var source1 =
+@"
+public partial interface I1<T> where T : I1<T>
+{
+    virtual static " + type + " operator " + op1 + " " + paramList + @" => throw null;
+    virtual static " + type + " operator " + op2 + " " + paramList + @" => throw null;
+}
+";
+
+            return source1;
+        }
+
+        private static string InParameter_03_Source2(string type, string op1, string op2, string paramList)
+        {
+            var source2 =
+@"
+partial class C1 : I1<C1>
+{
+    static " + type + " I1<C1>.operator " + op1 + " " + paramList + @" => throw null;
+    static " + type + " I1<C1>.operator " + op2 + " " + paramList + @" => throw null;
+}
+";
+
+            return source2.Replace("T", "C1");
+        }
+
+        [WorkItem("https://github.com/dotnet/roslyn/issues/68767")]
+        [Theory]
+        [InlineData("I1<T>", ">", "<", "(I1<T> x, in I1<T> y)")]
+        [InlineData("I1<T>", ">=", "<=", "(I1<T> x, in I1<T> y)")]
+        [InlineData("T", "==", "!=", "(T x, in T y)")]
+        public void InParameters_04(string type, string op1, string op2, string paramList)
+        {
+            string source2 = InParameter_03_Source2(type, op1, op2, paramList);
+            InParameter_02_Validate(InParameter_03_Source1_Abstract(type, op1, op2, paramList), source2);
+            InParameter_02_Validate(InParameter_03_Source1_Virtual(type, op1, op2, paramList), source2);
+        }
+
+        [WorkItem("https://github.com/dotnet/roslyn/issues/68767")]
+        [Fact]
+        public void InParameters_05()
+        {
+            var source1_Abstract =
+@"
+public interface I1<T> where T : I1<T>
+{
+    abstract static implicit operator int(in T x);
+    abstract static explicit operator T(in int x);
+    abstract static explicit operator checked T(in int x);
+}
+";
+            var source1_Virtual =
+@"
+public interface I1<T> where T : I1<T>
+{
+    virtual static implicit operator int(in T x) => throw null;
+    virtual static explicit operator T(in int x) => throw null;
+    virtual static explicit operator checked T(in int x) => throw null;
+}
+";
+
+            var source2 =
+@"
+partial class C1 : I1<C1>
+{
+    static implicit I1<C1>.operator int(in C1 x) => throw null;
+    static explicit I1<C1>.operator C1(in int x) => throw null;
+    static explicit I1<C1>.operator checked C1(in int x) => throw null;
+}
+";
+
+            InParameter_01_Validate(source1_Abstract, source2);
+            InParameter_01_Validate(source1_Virtual, source2);
+        }
     }
 }

